vote-circuit/src/lib.rs

```diff
@@ -0,0 +1,460 @@
+                MAX_MERKLE_DEPTH
+            );
+        }
+
+        // Validate merkle proof length
+        if self.private_inputs.merkle_siblings.len() != self.private_inputs.path_indices.len() {
+            bail!(
+                "Merkle proof length mismatch: {} siblings vs {} path indices",
+                self.private_inputs.merkle_siblings.len(),
+                self.private_inputs.path_indices.len()
+            );
+        }
+
+        // Helper to set HashOutTarget from [F; 4]
+        fn set_hash_target_witness_from_felts(
+            pw: &mut PartialWitness<F>,
+            target: HashOutTarget,
+            val: &[F; 4],
+        ) -> anyhow::Result<()> {
+            pw.set_hash_target(target, HashOut { elements: *val })?;
+            Ok(())
+        }
+
+        // Set public input witnesses
+        set_hash_target_witness_from_felts(
+            pw,
+            targets.proposal_id,
+            &self.public_inputs.proposal_id,
+        )?;
+        set_hash_target_witness_from_felts(
+            pw,
+            targets.expected_merkle_root,
+            &self.public_inputs.merkle_root,
+        )?;
+        pw.set_bool_target(targets.vote, self.public_inputs.vote)?;
+        set_hash_target_witness_from_felts(
+            pw,
+            targets.expected_nullifier,
+            &self.public_inputs.nullifier,
+        )?;
+
+        // Set private input witnesses
+        set_hash_target_witness_from_felts(
+            pw,
+            targets.private_key,
+            &self.private_inputs.private_key,
+        )?;
+        pw.set_target(
+            targets.actual_merkle_depth,
+            F::from_canonical_usize(self.private_inputs.actual_merkle_depth),
+        )?;
+
+        for i in 0..MAX_MERKLE_DEPTH {
+            if i < self.private_inputs.actual_merkle_depth {
+                set_hash_target_witness_from_felts(
+                    pw,
+                    targets.merkle_siblings[i],
+                    &self.private_inputs.merkle_siblings[i],
+                )?;
+                pw.set_bool_target(targets.path_indices[i], self.private_inputs.path_indices[i])?;
+            } else {
+                let zero_felts = [F::ZERO; 4];
+                set_hash_target_witness_from_felts(pw, targets.merkle_siblings[i], &zero_felts)?;
+                pw.set_bool_target(targets.path_indices[i], false)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use circuit_common::circuit::C;
+    use plonky2::{
+        field::types::Field,
+        hash::poseidon::PoseidonHash,
+        iop::witness::PartialWitness,
+        plonk::{circuit_data::CircuitConfig, config::Hasher},
+    };
+
+    fn bytes_to_felts(bytes: &[u8; 32]) -> [F; 4] {
+        let mut felts = [F::ZERO; 4];
+        for (i, chunk) in bytes.chunks(8).enumerate() {
+            felts[i] = F::from_canonical_u64(u64::from_le_bytes(chunk.try_into().unwrap()));
+        }
+        felts
+    }
```

Comment:
Don't we already have this function in `utils.rs`? Let's move the whole utility file to the common circuit crate too